File: LeetCode-POTD-Solutions/Jump-Game-III.cpp
1class Solution {
2public:
3    bool canReach(vector<int>& arr, int start) {
4        int n = arr.size();
5        vector<bool> visited(n, false);
6        return dfs(arr, start, visited);
7    }
8
9private:
10    bool dfs(vector<int>& arr, int i, vector<bool>& visited) {
11        // Out of bounds or already visited
12        if (i < 0 || i >= arr.size() || visited[i]) {
13            return false;
14        }
15
16        // Found a zero
17        if (arr[i] == 0) {
18            return true;
19        }
20
21        visited[i] = true;
22
23        // Try both possible jumps
24        return dfs(arr, i + arr[i], visited) ||
25               dfs(arr, i - arr[i], visited);
26    }
27};
28

File: LeetCode-POTD-Solutions/Jump Game III.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        boolean[] visited = new boolean[arr.length];
        return dfs(arr, start, visited);
    }

    private boolean dfs(int[] arr, int idx, boolean[] visited) {
        // Out of bounds or already visited
        if (idx < 0 || idx >= arr.length || visited[idx]) {
            return false;
        }

        // Found a zero
        if (arr[idx] == 0) {
            return true;
        }

        visited[idx] = true;

        // Jump forward or backward
        return dfs(arr, idx + arr[idx], visited) ||
               dfs(arr, idx - arr[idx], visited);
    }
}